python/gauss2d/fit/pybind11.cc

```diff
@@ -0,0 +1,287 @@
+/*
+ * This file is part of gauss2d.
+ *
+ * Developed for the LSST Data Management System.
+ * This product includes software developed by the LSST Project
+ * (https://www.lsst.org).
+ * See the COPYRIGHT file at the top-level directory of this distribution
+ * for details of code ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+//#include <iostream>
+#include <limits>
+#include <memory>
+#include <string>
+
+#include "gauss2d/centroid.h"
+#include "gauss2d/ellipse.h"
+#include "gauss2d/fit/centroidparameters.h"
+#include "gauss2d/fit/ellipseparameters.h"
+#include "gauss2d/fit/parameters.h"
+#include "gauss2d/fit/transforms.h"
+#include "parameters/transform.h"
+#include "parameters/unit.h"
+
+#include <pybind11/attr.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+//#include <pybind11/stl_bind.h>
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+// This is so annoying that I'm considering borrowing a solution for static string 
+template<typename T>
+constexpr std::string_view suffix_type();
+
+template<> constexpr std::string_view suffix_type<double>() { return "D";}
+
+template<typename T>
+std::string suffix_type_str() { return std::string(suffix_type<T>()); }
+
+template<typename T>
+constexpr std::string_view limits_name();
+
+template<> constexpr std::string_view limits_name<double>() { return ", gauss2d.fit.LimitsD";}
+
+// Yes, this class exists solely to hold a string so that it can be initialized first
+// in an inheritance hierarchy. Solve this better if you can.
```

Comment:
Limits needs to be a literal type to be used as a `constexpr` back in `parameters`, so it can't store a non-literal `std::string`. That's for each Parameter to have a quasi-singleton `static const Limits<T> limits_maximal` (regardless of whether it uses CRTP or not). This was the only solution I found that worked. The `Name` class is needed to ensure that the string is initialized first; a string member would be initialized after the `parameters::Limits<T>` and compile but segfault.

All of this would be trivial with `constexpr std::string`, but that is sadly a C++20 feature.

Alternatively, I could simply drop the naming of `Limits` altogether, but I think it would add user-friendliness to have names (e.g. naming position limits based on bboxes, giving more restrictive limits on Sersic indices a name like `Default` or `GaussianMixtureDefault`, etc.).